### PR TITLE
[Feat] find_or_insert and assign APIs

### PR DESCRIPTION
**Pull request by jiashuy**
[Feat]: find_or_insert and assign APIs.

### CODE REVIEW

File: include/merlin/core_kernels.cuh [hunk 1]
```diff
@@ -1966,5 +1968,669 @@ __global__ void dump_kernel(const Table<K, V, M>* __restrict table,
...
+  for (int i = 0; i < n_worker; i++) {
+    thds[i].join();
+  }
+}
+
+/* find or insert with the end-user specified meta.
+ */
+template <class K, class V, class M, uint32_t TILE_SIZE = 4>
+__global__ void find_or_insert_kernel(const Table<K, V, M>* __restrict table,
+                              const K* __restrict keys, V** __restrict vectors,
```
Comment (rhdong) [inline include/merlin/core_kernels.cuh:2016]: Need to format by following the guide: https://github.com/NVIDIA-Merlin/HierarchicalKV/blob/master/STYLE_GUIDE.md

**Comment by rhdong** [comment]
Hi @jiashuy , can you collect the benchmark data for these two APIs and update it to the relative section in the [Readme](https://github.com/NVIDIA-Merlin/HierarchicalKV#benchmark--performancewip), thank you so much!

**Comment by rhdong** [comment]
Hello @WonderingWJ , could you help review this PR for Jiashu, thank you~

File: include/merlin/core_kernels.cuh (hunk 1 above)
Comment (jiashuy) [reply include/merlin/core_kernels.cuh:2016]: OK, I will do it

**Comment by jiashuy** [comment]
> 



>

**Comment by jiashuy** [comment]
> Hi @jiashuy , can you collect the benchmark data for these two APIs and update it to the relative section in the [Readme](https://github.com/NVIDIA-Merlin/HierarchicalKV#benchmark--performancewip), thank you so much!

OK, no problem

File: include/merlin/core_kernels.cuh (hunk 1 above)
Comment (jiashuy) [reply include/merlin/core_kernels.cuh:2016]: Nice tool! I learned a lot. Thank you.(^▽^ )

File: tests/googletest
Comment (rhdong) [inline tests/googletest:1]: should not modify the googletest.

File: include/merlin_hashtable.cuh [hunk 2]
```diff
@@ -575,6 +575,226 @@ class HashTable {
     CudaCheckError();
   }
 
+  /**
+   * @brief Searches the hash table for the specified keys.
+   *
+   * @note When a key is missing, the value in @p values and @p metas will be inserted.
```
Comment (rhdong) [inline include/merlin_hashtable.cuh:581]: Please append the note to brief.

File: include/merlin/core_kernels.cuh [hunk 3]
```diff
@@ -1966,5 +1968,669 @@ __global__ void dump_kernel(const Table<K, V, M>* __restrict table,
...
+
+  auto functor = [=](V** __restrict table_value_addrs, V* __restrict param_values,
+                    const int* __restrict offset, int handled_size,
+                    int trunk_size) -> void {
+    for (int i = handled_size; i < handled_size + trunk_size; i++) {
+      if (founds[offset[i]] && table_value_addrs[i] != nullptr) {
+        memcpy(param_values + offset[i] * dim, table_value_addrs[i], sizeof(V) * dim);
+      }
+      else {
+        memcpy(table_value_addrs[i], param_values + offset[i] * dim, sizeof(V) * dim);
```
Comment (rhdong) [inline include/merlin/core_kernels.cuh:1990]: If need to check `table_value_addrs[i] != nullptr` here?

File: include/merlin/core_kernels.cuh [hunk 4]
```diff
@@ -1966,5 +1968,669 @@ __global__ void dump_kernel(const Table<K, V, M>* __restrict table,
...
+      }
+    }
+  };
+
+  size_t trunk_size = N / n_worker;
```
Comment (rhdong) [inline include/merlin/core_kernels.cuh:1995]: If `N < n_worker` is true, trunk_size will be `0`.

File: include/merlin/core_kernels.cuh [hunk 5]
```diff
@@ -1966,5 +1968,669 @@ __global__ void dump_kernel(const Table<K, V, M>* __restrict table,
...
+    auto const found_vote = find_in_bucket<K, V, M, TILE_SIZE>(
+        g, bucket->keys, update_key, tile_offset, start_idx, bucket_max_size);
+
+    if (found_vote) {
+      const int dst_lane = __ffs(found_vote) - 1;
+      key_pos = (start_idx + tile_offset + dst_lane) & (bucket_max_size - 1);
+      V* dst = bucket->vectors + key_pos * dim;
+
+      if (rank == dst_lane) {
+        if (metas != nullptr) {
```
Comment (rhdong) [inline include/merlin/core_kernels.cuh:2259]: Why not use `update_meta` directly?

File: include/merlin/core_kernels.cuh [hunk 6]
```diff
@@ -1966,5 +1968,669 @@ __global__ void dump_kernel(const Table<K, V, M>* __restrict table,
...
+          bucket->metas[key_pos].store(metas[key_idx], cuda::std::memory_order_relaxed);
+        }
+        else {
+          M cur_meta = bucket->cur_meta.fetch_add(1) + 1;
+          bucket->metas[key_pos].store(cur_meta, cuda::std::memory_order_relaxed);
+        }
```
Comment (rhdong) [inline include/merlin/core_kernels.cuh:2265]: Why not use `update_meta` kernel directly?

File: include/merlin/core_kernels.cuh [hunk 7]
```diff
@@ -1966,5 +1968,669 @@ __global__ void dump_kernel(const Table<K, V, M>* __restrict table,
...
+      }
+      if (local_size >= bucket_max_size) {
+        refresh_bucket_meta<K, V, M, TILE_SIZE>(g, bucket, bucket_max_size);
+      }
+
+      lock<Mutex, TILE_SIZE, true>(g, table->locks[bkt_idx]);
+      copy_vector<V, TILE_SIZE>(g, update_value, dst, dim);
+      unlock<Mutex, TILE_SIZE, true>(g, table->locks[bkt_idx]);
+    }
+    ///TODO:(jiashuy) conform: otherwise ignore it
```
Comment (rhdong) [inline include/merlin/core_kernels.cuh:2275]: Is this `TODO` needed?  May I miss some thing?

File: tests/test_util.cuh [hunk 8]
```diff
@@ -30,6 +30,10 @@
 #include "merlin/utils.cuh"
 #include "merlin_hashtable.cuh"
 
+#define UNEQUAL_EXPR(expr1, expr2)                            \
```
Comment (rhdong) [inline tests/test_util.cuh:33]: I perfer this( sorry for no formatted) :
`#define UNEQUAL_EXPR(expr1, expr2)                            \`
`  {std::cout <<  __FILE__ << ":" << __LINE__ << ":Unequal\n"   \`
`         << "\t\t" << #expr1 << " != " << #expr2 << "\n"; }`

File: tests/merlin_hashtable_test.cc.cu [hunk 9]
```diff
@@ -1944,6 +1944,367 @@ void test_evict_strategy_customized_correct_rate(size_t max_hbm_for_vectors,
...
+      size_t max_hbm_for_vectors, 
+      bool use_constant_memory, 
+      const uint64_t init_size_in_mb, 
+      const uint64_t max_size_in_mb,
+      const uint64_t THREAD_N,
+      const uint64_t BATCH_0_SIZE,
+      const uint64_t BATCH_1_SIZE,
+      bool capacity_silent = true
+    ) {
+  const uint64_t INIT_CAPACITY = nv::merlin::MB(init_size_in_mb);
```
Comment (rhdong) [inline tests/merlin_hashtable_test.cc.cu:1958]: As I mentioned in the meeting, the unit of capacity is key-value pairs, not `bytes`.

File: tests/merlin_hashtable_test.cc.cu [hunk 10]
```diff
@@ -1944,6 +1944,367 @@ void test_evict_strategy_customized_correct_rate(size_t max_hbm_for_vectors,
 
   CudaCheckError();
 }
+
+void test_insert_or_assign_multi_threads(
+      size_t max_hbm_for_vectors, 
+      bool use_constant_memory,
```
Comment (rhdong) [inline tests/merlin_hashtable_test.cc.cu:1950]: The constant memory configuration has been removed, please rebase on master and process the conflicts, thanks.

File: tests/test_util.cuh [hunk 11]
```diff
@@ -111,6 +115,31 @@ void create_random_keys(K* h_keys, M* h_metas, int KEY_NUM) {
   }
 }
 
+template <class K, class M, class V, size_t DIM = 16>
+void create_random_keys(K* h_keys, M* h_metas, V* h_vectors, int KEY_NUM) {
```
Comment (rhdong) [inline tests/test_util.cuh:119]: Should remove the same func from `merlin_hashtable_test.cc.cu`.

File: tests/test_util.cuh [hunk 12]
```diff
@@ -121,6 +150,57 @@ inline uint64_t Murmur3HashHost(const uint64_t& key) {
   return k;
 }
 
+template <class K, class M, class V, size_t DIM = 16>
+void create_continuous_keys(K* h_keys, M* h_metas, V* h_vectors, int KEY_NUM,
```
Comment (rhdong) [inline tests/test_util.cuh:154]: Should remove the same func from `merlin_hashtable_test.cc.cu` and please notice the `DIM`.

File: tests/test_util.cuh [hunk 13]
```diff
@@ -121,6 +150,57 @@ inline uint64_t Murmur3HashHost(const uint64_t& key) {
...
+    if (h_vectors != nullptr) {
+      for (size_t j = 0; j < DIM; j++) {
+        h_vectors[i * DIM + j] = static_cast<float>(h_keys[i] * 0.00001);
+      }
+    }
+  }
+}
+
+template <class K, class M, class V, size_t DIM = 16>
+void create_keys_in_one_buckets(K* h_keys, M* h_metas, V* h_vectors,
```
Comment (rhdong) [inline tests/test_util.cuh:168]: Should remove the same func from `merlin_hashtable_test.cc.cu` and please notice to process the `DIM` properly.

File: tests/find_or_insert_test.cc.cu [hunk 14]
```diff
@@ -0,0 +1,2309 @@
...
+                                               const K& pattern,
+                                               const M& threshold) {
+  return meta > threshold;
+}
+
+template <class K, class M>
+__device__ Table::Pred ExportIfPred = export_if_pred<K, M>;
+
+
+void test_basic_mi(size_t max_hbm_for_vectors, bool use_constant_memory) {
```
Comment (rhdong) [inline tests/find_or_insert_test.cc.cu:62]: If everything has `mi`, the `mi` would be unnecessary. And now you make a seperate test file.

Comment (jiashuy) [reply tests/find_or_insert_test.cc.cu:62]: OK, I get

File: include/merlin/core_kernels.cuh (hunk 3 above)
Comment (jiashuy) [reply include/merlin/core_kernels.cuh:1990]: Yes, For correct usage it needs, and I will check it here. But in context, the table_value_addrs[i] should not be a nullptr, as both find or insert branch will give it a legal value.(It is only used by find_or_insert operation now)

File: tests/googletest
Comment (jiashuy) [reply tests/googletest:1]: Sorry, I will remove the googletest submodule from the git's working-storage section

File: include/merlin/core_kernels.cuh (hunk 4 above)
Comment (jiashuy) [reply include/merlin/core_kernels.cuh:1995]: OK, I will check before launch a thread to avoid its overhead

Comment (jiashuy) [reply include/merlin/core_kernels.cuh:1995]: And try to make every `worker`'s workload balanced

File: include/merlin/core_kernels.cuh (hunk 6 above)
Comment (jiashuy) [reply include/merlin/core_kernels.cuh:2265]: OK, I will use `update_meta`. Forgive me forgotten it. :D

File: tests/test_util.cuh (hunk 8 above)
Comment (jiashuy) [reply tests/test_util.cuh:33]: OK, I will update. (^▽^ )

File: include/merlin/core_kernels.cuh (hunk 3 above)
Comment (rhdong) [reply include/merlin/core_kernels.cuh]: Make sense！

File: include/merlin_hashtable.cuh [hunk 15]
```diff
@@ -579,6 +579,230 @@ class HashTable {
...
+   * @param values The values to search on GPU-accessible memory with
+   * shape (n, DIM).
+   * @param metas The metas to search on GPU-accessible memory with shape (n).
+   * @parblock
+   * If @p metas is `nullptr`, the meta for each key will not be returned.
+   * @endparblock
+   * @param stream The CUDA stream that is used to execute the operation.
+   *
+   */
+  void find_or_insert(const size_type n, const key_type* keys,  // (n)
```
Comment (rhdong) [inline include/merlin_hashtable.cuh:598]: Hi @WonderingWJ , please help confirm the API define is matching your requirement, thank you~

File: include/merlin_hashtable.cuh [hunk 16]
```diff
@@ -579,6 +579,230 @@ class HashTable {
...
+   * such as the timestamp of the key insertion, number of the key
+   * occurrences, or another value to perform a custom eviction strategy.
+   *
+   * The @p metas should be `nullptr`, when the LRU eviction strategy is
+   * applied.
+   * @endparblock
+   *
+   * @param stream The CUDA stream that is used to execute the operation.
+   */
+  void assign(const size_type n,
```
Comment (rhdong) [inline include/merlin_hashtable.cuh:723]: Hi @WonderingWJ , please help confirm the API define is matching your requirement, thank you~

**Review by WonderingWJ — APPROVED** [review]
Test passed

**Review by WonderingWJ — DISMISSED** [review]

**Review by rhdong — DISMISSED** [review]
LGTM

**Review by rhdong — APPROVED** [review]
LGTM

**Pull request merged**